File: Decorator/WeaponSample/src/WeaponAttachment.java
public class WeaponAttachment implements IWeapon {
    protected IWeapon weapon;
    protected String name;
    public WeaponAttachment(IWeapon weapon) {
        this.weapon = weapon;
    }

    @Override
    public void Fire() {
        weapon.Fire();
    }

    @Override
    public String GetName() {
        return this.name;
    }

    @Override
    public void SetName(String name) {
        this.name = name;
    }
}

File: Decorator/csharp/Sample/Decorator/Weapon/WeaponAttachment.cs
namespace Decorator.Weapon
{
    public abstract class WeaponAttachment : IWeapon
    {
        protected IWeapon _weapon;

        protected WeaponAttachment(IWeapon weapon)
        {
            _weapon = weapon;
        }

        public virtual void Fire()
        {
            _weapon.Fire();
        }

        public string Name { get; }
    }

}